src/Microsoft.ApplicationInsights/Extensibility/Implementation/CorrelationLookup/FailedRequestsManager.cs

```diff
@@ -0,0 +1,124 @@
+
+        private ConcurrentDictionary<string, FailedResult> failingInstrumentationKeys = new ConcurrentDictionary<string, FailedResult>();
+
+        internal FailedRequestsManager(int retryWaitTimeSeconds = 30)
+        {
+            this.retryWaitTimeSeconds = TimeSpan.FromSeconds(retryWaitTimeSeconds);
+        }
+
+        /// <summary>
+        /// FetchAppIdFromService failed.
+        /// Registers failure for further action in future.
+        /// </summary>
+        /// <param name="instrumentationKey">Instrumentation key for which the failure occurred.</param>
+        /// <param name="httpStatusCode">Response code from AppId Endpoint.</param>
+        public void RegisterFetchFailure(string instrumentationKey, HttpStatusCode httpStatusCode)
+        {
+            this.failingInstrumentationKeys.TryAdd(instrumentationKey, new FailedResult(this.retryWaitTimeSeconds, httpStatusCode));
+
+            CorrelationLookupEventSource.Log.FetchAppIdFailedWithResponseCode(httpStatusCode.ToString());
+        }
+
+        /// <summary>
+        /// Registers failure for further action in future.
+        /// </summary>
+        /// <param name="instrumentationKey">Instrumentation key for which the failure occurred.</param>
+        /// <param name="ex">Exception indicating failure.</param>
+        public void RegisterFetchFailure(string instrumentationKey, Exception ex)
+        {
+            if (ex is AggregateException ae)
+            {
+                var innerException = ae.Flatten().InnerException;
+                if (innerException != null)
+                {
+                    this.RegisterFetchFailure(instrumentationKey, innerException);
+                    return;
+                }
+            }
+            else if (ex is WebException webException && webException.Response != null && webException.Response is HttpWebResponse httpWebResponse)
+            {
+                this.failingInstrumentationKeys.TryAdd(instrumentationKey, new FailedResult(this.retryWaitTimeSeconds, httpWebResponse.StatusCode));
+            }
+            else
+            {
+                this.failingInstrumentationKeys.TryAdd(instrumentationKey, new FailedResult(this.retryWaitTimeSeconds));
+            }
+
+            CorrelationLookupEventSource.Log.FetchAppIdFailed(this.GetExceptionDetailString(ex));
+        }
+
+        public bool CanRetry(string instrumentationKey)
+        {
+            if (this.failingInstrumentationKeys.TryGetValue(instrumentationKey, out FailedResult lastFailedResult))
+            {
+                if (lastFailedResult.CanRetry())
+                {
+                    this.failingInstrumentationKeys.TryRemove(instrumentationKey, out FailedResult value);
+                    return true;
+                }
+                else
+                {
+                    return false;
+                }
+            }
+
+            Debug.WriteLine("no failed result found");
```

Comment:
we do not do ` Debug.WriteLine` in our SDK. If you need it - create an verbose EventSource message. If not - please remove

Reply:
Thanks for the clarification.
I need it right now while i'm investigating a bug in one of the failing unit tests.
I'll be careful to remove all uses.